feat: run async read on std::thread instead of uv threadpool

Julusian opened the pull request:
This resolves the issue with not being able to read from more than 4 devices simultaneously unless `UV_THREADPOOL_SIZE` is increased.
https://github.com/node-hid/node-hid/issues/408

Instead of using the uv-pool to run the reads, a normal c++ thread is started to read from each device when `resume()` is first called or a listener is added. 

I have tried to keep functionality the same, but it is possible there are some subtle differences.
The change ended up being a bit bigger than I was hoping, to get the lifetime of the device to behave the same as before.

One thing that has changed is the removal of:
https://github.com/node-hid/node-hid#devicereadcallback
I could bring it back, but it would behave differently when called for a second time

Some backround reading: https://nodejs.org/en/docs/guides/dont-block-the-event-loop/

Julusian commented:
So far I have tested it on linux (ubuntu 20.04) with libusb. I dont see any risk of it having broken other platforms as long as it compiles. I shall try it on more to verify when I can

Julusian commented:
I've been running this for a couple of weeks on macos too, with no noticable issues

nytamin commented:
This seems to run well on Windows too. Tested with with 3 Streamdecks and 1 X-keys, all seem to work fine with this.
(I came from node-hid `2.1.1` which did not.)

Julusian commented:
@todbot any chance on getting this merged soon?

Julusian commented:
I have updated the readme, and replaced any uses of `.read()` in the examples. I havent tested them all, but did give the ps3 ones a quick test as I have a compatible controller.

If this is going to form the basis of a 3.0.0 release, then it would be worth thinking about whether there are other breaking changes that would be good to do at the same time.

Julusian commented:
I was thinking that it would be nice to see some more methods converted to being async/promise based. But I realise it would be a better approach to add them as new methods, and mark the old ones as deprecated for a while first. I dont see any technical reason why we can't support both for a while.

But one thing that might be good to do at the same time as removing `read()` is removing `readSync()` and `readTimeout()`. They both block the main thread while executing, meaning nothing else can run (not even timers). But if they were promise/callback based with the work being done on some background thread then things would be smoother. 
I can see some value in them or experimentation or tiny applications, but if used in an application of a reasonable size those methods will cause performance problems.
So ideally I would like to see them removed, and I guess a new `readAsync(timeout?: number | undefined): Promise<Buffer | undefined>` (perhaps with a better name) added as a safer(non-blocking) replacement.

Julusian commented:
After playing around with making some async methods, I am starting to lean towards wanting to replace the whole HID class with an async version. That could be done as a new class added alongside the existing one in v2.x, or it could replacing the current HID class in v3.x.
I am up for discussion on that, but I am thinking that the most manageable way to do this is to do a separate class, otherwise an extra mutex is needed to ensure that the sync and async methods arent used at the same point in time. And the HID.cc file is already rather large.

What are your thoughts on whether both sync and async methods should be supported?
And does an almost complete rewrite for v3.x sound like a reasonable plan? (I seem to be making a habit of this :smile:  #403 )

If we want to keep the sync methods around for longer, then perhaps this pr should be done for just the async class for now. That avoids a semver major change. I do think that the sync methods should be marked as deprecated with a plan for fully removing them, if they are kept around. Long term I dont think its a good idea to keep both sync and async versions around, it will double the amount of code and will make it easy for libraries to use the inefficient sync methods, without consumers being aware.

--------

I did some crude and unscientific performance tests. It really needs to be repeated with smaller buffers too, and on other platforms than just linux on my laptop. hidraw and libusb had similar numbers.

With a 1kb buffer, `hid.write()` takes around 1700ms for 10000 samples.
My initial `hid.writeAsync()` takes around 1500ms for 10000 samples.

`hid.sendFeatureReportAsync()` is proving less beneficial with a 32byte payload, being 10-100ms slower than `hid.sendFeatureReport()` which was taking ~1050ms. But still, that is a small cost and has the benefit of being async.
Hacking in some timers reveals that 900ms of 1200ms is being spent in `hid_send_feature_report`, so by doing it async, we are giving the main thread back a lot of time. 

I have implemented a fairly simple jobQueue, to ensure that it isnt doing concurrent operations from different threads. Making use of the uv pool for now.
The async doesnt feel like a big gain based on the numbers, but the fact that 900 of 1200ms is being done on a background thread instead, means user code has a lot more cpu time available to it. Or in an application using 4 hid devices, that means that the `hid_send_feature_report` will be able to run in parallel.

Julusian commented:
Yes I am aware of that. I have made some assumptions about what I expect will be safe, but I do need to verify those assumptions. I expect some of it to hold true, as I think some places use this in worker-threads which has multiple main-threads, and I havent heard of any weird crashes there (yet)

But if they dont hold true, then I can create a single workqueue to be shared across all hid operations, rather than a queue per opened device. It will limit performance in setups with multiple devices, but wont completely negate the benefit. That 900 of 1200ms being done on a background thread will still give user code more time to run on the main thread.

I do need to do research into this, and verify if even the way that reads are currently being done is safe

Julusian commented:
I've spent some time reading through the hidapi source. I skimmed over many blocks, and focused on checking my assumptions, and have concluded the following:
* Doing any other operation in parallel with a read is not thread safe. they both modify the same 'last_error' property on some platforms. From what I can see that is the only non-thread safe bit about doing those operations concurrently, but does mean we should probably stop doing it (or comment out where it sets the error strings so that we dont have that issue)
* Doing hid_open/hid_enumerate/hid_close calls on multiple threads on hidraw can leak an error string. It uses a thread_local pointer, with no way to clear it. The same problem has been flagged for a pr for macos too https://github.com/libusb/hidapi/pull/314#discussion_r687541912. This is a problem in node-hid already, if you use it in a worker thread. Ultimately it is unlikely to leak more than 100bytes, so the leak will not be noticable except in some very extreme and unusual circumstances.
* It dont see any reason to suggest that having a thread per device is unsafe. Some old issues suggest that this model is safe https://github.com/signal11/hidapi/issues/72#issuecomment-7114686. This is assuming that the underlying calls made are threadsafe, which from a quick google sounds like they all are.

I made sure to read all 4 backends (hidraw, libusb, mac and windows)

From this I might need to slightly revise my plan, as I am uncertain if we should keep the current setup of separate read thread per device. But the idea of having a thread for the rest of the operations is still fine. 
It looks like the only truely thread-safe way to do this is to use hidapi in nonblocking mode and poll. But I think that doing that is going to be complex, and will likely lead to complaints about node-hid becoming a battery hog or feel sluggish (depending on how often the polling is performed, and how many threads are doing the polling). 
So I am a little tempted to ignore the non-safety from when it reports an error. even though I know that is a bad strategy.. The best solution I can think of is to use a modified version of hidapi that has the generation of error strings disabled. We arent using them anyway, and its only a few lines to comment out so wont cause big merge conflicts going forward.

Having a single thread doing the enumeration/opening of devices looks to be fine. In fact, this will be more thread-safe than it is now as with worker threads it is possible to execute the hid_enumerate concurrently. So I shall make a single 'application wide' thread/queue to process these operations.

cwilling commented:
I just found this while investigating why setting UV_THREADPOOL_SIZE works well enough on Linux to increase the number of simultaneous devices but makes no difference on Windows.

Building from source with this PR now enables 9 physical devices (8x X-Keys & 1x Elgato Streamdeck) to run simultaneously on Windows10 so I'm highly impressed - also relieved that a solution is in the air. Great work @Julusian !

I hope a version of this can make it to the npm publish stage soon.

Julusian commented:
Replaced by #474 which includes this among its many other changes
